{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c589d147",
   "metadata": {},
   "source": [
    "We set some variables within our environment for easy access to settings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "421ef978",
   "metadata": {},
   "outputs": [],
   "source": [
    "%env THREADS 12"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4feb9e46",
   "metadata": {},
   "source": [
    "We assume conda is installed properly and the user has write access to create and run an environment.\n",
    "\n",
    "We also assume enough available resources for the run. The major one is 64gb or memory. \n",
    "\n",
    "The rest of the parameter files assume a 12 thread CPU (the one we had for the project).\n",
    "\n",
    "We setup an environment and apply some must needed fixes to all the applicable software:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32ef79ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 'defaults' already in 'channels' list, moving to the top\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Warning: 'bioconda' already in 'channels' list, moving to the top\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Warning: 'conda-forge' already in 'channels' list, moving to the top\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Channels:\n",
      " - conda-forge\n",
      " - bioconda\n",
      " - defaults\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: failed\n",
      "\n",
      "LibMambaUnsatisfiableError: Encountered problems while solving:\n",
      "  - package pygtftk-1.0.1-py36hebb334c_0 requires python >=3.6,<3.7.0a0, but none of the providers can be installed\n",
      "\n",
      "Could not solve for environment specs\n",
      "The following packages are incompatible\n",
      "├─ \u001b[32mpygtftk\u001b[0m is installable with the potential options\n",
      "│  ├─ \u001b[32mpygtftk [1.0.1|1.0.2|...|1.1.0]\u001b[0m would require\n",
      "│  │  └─ \u001b[32mpython >=3.6,<3.7.0a0 \u001b[0m, which can be installed;\n",
      "│  ├─ \u001b[32mpygtftk 1.1.0\u001b[0m would require\n",
      "│  │  └─ \u001b[32mpython >=3.7,<3.8.0a0 \u001b[0m, which can be installed;\n",
      "│  ├─ \u001b[32mpygtftk [1.1.1|1.1.2|...|1.2.7]\u001b[0m would require\n",
      "│  │  ├─ \u001b[32mpython >=3.6,<3.7.0a0 \u001b[0m, which can be installed;\n",
      "│  │  └─ \u001b[32mpython_abi 3.6.* *_cp36m\u001b[0m, which can be installed;\n",
      "│  ├─ \u001b[32mpygtftk [1.1.1|1.1.2|...|1.2.7]\u001b[0m would require\n",
      "│  │  ├─ \u001b[32mpython >=3.7,<3.8.0a0 \u001b[0m, which can be installed;\n",
      "│  │  └─ \u001b[32mpython_abi 3.7.* *_cp37m\u001b[0m, which can be installed;\n",
      "│  ├─ \u001b[32mpygtftk [1.3.0|1.4.0|1.5.0|1.5.1|1.5.3]\u001b[0m would require\n",
      "│  │  ├─ \u001b[32mpython >=3.8,<3.9.0a0 \u001b[0m, which can be installed;\n",
      "│  │  └─ \u001b[32mpython_abi 3.8.* *_cp38\u001b[0m, which can be installed;\n",
      "│  └─ \u001b[32mpygtftk [1.6.0|1.6.1|1.6.2]\u001b[0m would require\n",
      "│     ├─ \u001b[32mpython >=3.9,<3.10.0a0 \u001b[0m, which can be installed;\n",
      "│     └─ \u001b[32mpython_abi 3.9.* *_cp39\u001b[0m, which can be installed;\n",
      "└─ \u001b[31mpython 3.11** \u001b[0m is not installable because there are no viable options\n",
      "   ├─ \u001b[31mpython [3.11.0|3.11.10|...|3.11.9]\u001b[0m conflicts with any installable versions previously reported;\n",
      "   └─ \u001b[31mpython [3.11.0|3.11.1|...|3.11.9]\u001b[0m would require\n",
      "      └─ \u001b[31mpython_abi 3.11.* *_cp311\u001b[0m, which conflicts with any installable versions previously reported.\n",
      "\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Collecting git+https://github.com/itmat/BEERS2\n",
      "  Cloning https://github.com/itmat/BEERS2 to /tmp/pip-req-build-d6hiw5c4\n",
      "  Running command git clone --filter=blob:none --quiet https://github.com/itmat/BEERS2 /tmp/pip-req-build-d6hiw5c4\n",
      "  Resolved https://github.com/itmat/BEERS2 to commit 0fdd2f7db38c1c6858aab415dc80a1e685afe788\n",
      "  Running command git submodule update --init --recursive -q\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting beers_utils@ git+https://github.com/itmat/BEERS_UTILS@52e1a00057d2939651ab587a4ad1a6873dc8d49e (from BEERS2==2.0)\n",
      "  Using cached beers_utils-0.2.0-py3-none-any.whl\n",
      "Collecting camparee@ git+https://github.com/itmat/CAMPAREE@1bb79db71ea8d74495334f1af74025624a4668c7 (from BEERS2==2.0)\n",
      "  Using cached camparee-0.4.2-py3-none-any.whl\n",
      "Requirement already satisfied: appdirs==1.4.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.4.4)\n",
      "Requirement already satisfied: attrs==22.2.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (22.2.0)\n",
      "Requirement already satisfied: boto3==1.26.85 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.26.85)\n",
      "Requirement already satisfied: botocore==1.29.85 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.29.85)\n",
      "Requirement already satisfied: certifi==2022.12.7 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2022.12.7)\n",
      "Requirement already satisfied: charset-normalizer==3.1.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.1.0)\n",
      "Requirement already satisfied: cloudpathlib==0.13.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.13.0)\n",
      "Requirement already satisfied: ConfigArgParse==1.5.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.5.3)\n",
      "Requirement already satisfied: connection-pool==0.0.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.0.3)\n",
      "Requirement already satisfied: contourpy==1.0.7 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.0.7)\n",
      "Requirement already satisfied: cycler==0.11.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.11.0)\n",
      "Requirement already satisfied: Cython==0.29.33 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.29.33)\n",
      "Requirement already satisfied: datrie==0.8.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.8.2)\n",
      "Requirement already satisfied: docutils==0.19 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.19)\n",
      "Requirement already satisfied: dpath==2.1.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.1.4)\n",
      "Requirement already satisfied: fastjsonschema==2.16.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.16.3)\n",
      "Requirement already satisfied: fonttools==4.39.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.39.0)\n",
      "Requirement already satisfied: gitdb==4.0.10 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.0.10)\n",
      "Requirement already satisfied: GitPython==3.1.31 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.1.31)\n",
      "Requirement already satisfied: humanfriendly==10.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (10.0)\n",
      "Requirement already satisfied: idna==3.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.4)\n",
      "Requirement already satisfied: Jinja2==3.1.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.1.2)\n",
      "Requirement already satisfied: jmespath==1.0.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.0.1)\n",
      "Requirement already satisfied: jsonschema==4.17.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.17.3)\n",
      "Requirement already satisfied: jupyter_core==5.2.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.2.0)\n",
      "Requirement already satisfied: kiwisolver==1.4.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.4.4)\n",
      "Requirement already satisfied: MarkupSafe==2.1.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.1.2)\n",
      "Requirement already satisfied: matplotlib==3.7.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.7.1)\n",
      "Requirement already satisfied: nbformat==5.7.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.7.3)\n",
      "Requirement already satisfied: numpy==1.24.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.24.2)\n",
      "Requirement already satisfied: packaging==23.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (23.0)\n",
      "Requirement already satisfied: pandas==1.5.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.5.3)\n",
      "Requirement already satisfied: Pillow==9.4.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (9.4.0)\n",
      "Requirement already satisfied: plac==1.3.5 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.3.5)\n",
      "Requirement already satisfied: platformdirs==3.1.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.1.0)\n",
      "Requirement already satisfied: prettytable==3.6.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.6.0)\n",
      "Requirement already satisfied: psutil==5.9.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.9.4)\n",
      "Requirement already satisfied: PuLP==2.7.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.7.0)\n",
      "Requirement already satisfied: pydantic==1.10.5 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.10.5)\n",
      "Requirement already satisfied: pyparsing==3.0.9 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.0.9)\n",
      "Requirement already satisfied: pyrsistent==0.19.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.19.3)\n",
      "Requirement already satisfied: pysam==0.20.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.20.0)\n",
      "Requirement already satisfied: python-dateutil==2.8.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.8.2)\n",
      "Requirement already satisfied: pytz==2022.7.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2022.7.1)\n",
      "Requirement already satisfied: PyYAML==6.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (6.0)\n",
      "Requirement already satisfied: requests==2.28.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.28.2)\n",
      "Requirement already satisfied: reretry==0.11.8 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.11.8)\n",
      "Requirement already satisfied: roman==4.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.0)\n",
      "Requirement already satisfied: s3transfer==0.6.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.6.0)\n",
      "Requirement already satisfied: scipy==1.10.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.10.1)\n",
      "Requirement already satisfied: six==1.16.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.16.0)\n",
      "Requirement already satisfied: smart-open==6.3.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (6.3.0)\n",
      "Requirement already satisfied: smmap==5.0.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.0.0)\n",
      "Requirement already satisfied: snakemake==7.24.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (7.24.0)\n",
      "Requirement already satisfied: stopit==1.1.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.1.2)\n",
      "Requirement already satisfied: tabulate==0.9.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.9.0)\n",
      "Requirement already satisfied: termcolor==2.2.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.2.0)\n",
      "Requirement already satisfied: throttler==1.2.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.2.2)\n",
      "Requirement already satisfied: toposort==1.10 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.10)\n",
      "Requirement already satisfied: traitlets==5.9.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.9.0)\n",
      "Requirement already satisfied: typing_extensions==4.5.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.5.0)\n",
      "Requirement already satisfied: urllib3==1.26.14 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.26.14)\n",
      "Requirement already satisfied: wcwidth==0.2.6 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.2.6)\n",
      "Requirement already satisfied: wrapt==1.15.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.15.0)\n",
      "Requirement already satisfied: yte==1.5.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.5.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "fatal: destination path 'L1EM' already exists and is not an empty directory.\n"
     ]
    }
   ],
   "source": [
    "# !conda create -n l1em python=3.11 openjdk=21 -y\n",
    "# !conda activate l1em\n",
    "%conda config --add channels defaults\n",
    "%conda config --prepend channels bioconda\n",
    "%conda config --prepend channels conda-forge\n",
    "%conda install -y python=3.11 openjdk=21 bwa samtools flux-simulator numpy scipy sra-tools trim-galore pysam bedtools pigz wget ipython=8.14 scikit-learn\n",
    "\n",
    "# Edit the flux-simulator caller script to overcome safety limitation due to old age coding standards (java code reflection)\n",
    "!if ! grep -q -- '--add-opens java.base/java.util=ALL-UNNAMED' $CONDA_PREFIX/share/flux-simulator-1.2.1-3/bin/flux-simulator; then sed -i '/^java -Xmx\\$FLUX_MEM.*/a --add-opens java.base/java.util=ALL-UNNAMED \\\\\\n--add-opens java.desktop/java.awt.font=ALL-UNNAMED \\\\\\n--add-opens java.base/java.text=ALL-UNNAMED' $CONDA_PREFIX/share/flux-simulator-1.2.1-3/bin/flux-simulator; fi\n",
    "\n",
    "# Install beers2 and camparee (which is automaticall installed with beers2 and must not be installed separetely)\n",
    "# %pip install git+https://github.com/itmat/BEERS2\n",
    "\n",
    "# Download the L1EM project from git\n",
    "!git clone https://github.com/geoleven/L1EM.git"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99f12b87",
   "metadata": {},
   "source": [
    "We go ahead and download the proper experiment data and the reference genomes, while also producing the relevant files from each download:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "3a53116e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Both FASTQ files already exist in the data folder. Skipping fasterq-dump.\n",
      "Trimmed files by trim_galore already exist. Skipping trim_galore.\n",
      "Humane genome 38 found. Assuming it is bwa indexed. Skipping downlaoding, unzipping and indexing it.\n",
      "Homo sapiens genome exists. Skipping downloading it.\n",
      "Indexing the Ensembl genome file...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 16.74 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=6199501436, availableWord=448219384\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 99999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 199999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 299999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 40 iterations done. 399999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 50 iterations done. 499999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 60 iterations done. 599999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 70 iterations done. 699999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 80 iterations done. 799999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 90 iterations done. 899999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 100 iterations done. 999999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 110 iterations done. 1099999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 120 iterations done. 1199999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 130 iterations done. 1299999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 140 iterations done. 1399999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 150 iterations done. 1499999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 160 iterations done. 1599999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 170 iterations done. 1699999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 180 iterations done. 1799999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 190 iterations done. 1899999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 200 iterations done. 1999999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 210 iterations done. 2099999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 220 iterations done. 2199999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 230 iterations done. 2299999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 240 iterations done. 2399999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 250 iterations done. 2499999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 260 iterations done. 2599999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 270 iterations done. 2699999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 280 iterations done. 2799999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 290 iterations done. 2899999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 300 iterations done. 2999999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 310 iterations done. 3099999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 320 iterations done. 3199999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 330 iterations done. 3299999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 340 iterations done. 3399999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 350 iterations done. 3499999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 360 iterations done. 3599999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 370 iterations done. 3699999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 380 iterations done. 3799999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 390 iterations done. 3899999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 400 iterations done. 3999999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 410 iterations done. 4099999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 420 iterations done. 4199999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 430 iterations done. 4299999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 440 iterations done. 4399999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 450 iterations done. 4499999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 460 iterations done. 4599999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 470 iterations done. 4699999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 480 iterations done. 4799999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 490 iterations done. 4899999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 500 iterations done. 4999999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 510 iterations done. 5099999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 520 iterations done. 5199999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 530 iterations done. 5299999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 540 iterations done. 5399999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 550 iterations done. 5499999996 characters processed.\n",
      "[BWTIncConstructFromPacked] 560 iterations done. 5596733548 characters processed.\n",
      "[BWTIncConstructFromPacked] 570 iterations done. 5683010028 characters processed.\n",
      "[BWTIncConstructFromPacked] 580 iterations done. 5759688796 characters processed.\n",
      "[BWTIncConstructFromPacked] 590 iterations done. 5827837068 characters processed.\n",
      "[BWTIncConstructFromPacked] 600 iterations done. 5888403388 characters processed.\n",
      "[BWTIncConstructFromPacked] 610 iterations done. 5942230876 characters processed.\n",
      "[BWTIncConstructFromPacked] 620 iterations done. 5990068876 characters processed.\n",
      "[BWTIncConstructFromPacked] 630 iterations done. 6032583388 characters processed.\n",
      "[BWTIncConstructFromPacked] 640 iterations done. 6070366396 characters processed.\n",
      "[BWTIncConstructFromPacked] 650 iterations done. 6103943980 characters processed.\n",
      "[BWTIncConstructFromPacked] 660 iterations done. 6133783772 characters processed.\n",
      "[BWTIncConstructFromPacked] 670 iterations done. 6160301452 characters processed.\n",
      "[BWTIncConstructFromPacked] 680 iterations done. 6183866396 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 688 iterations.\n",
      "[bwa_index] 2028.51 seconds elapse.\n",
      "[bwa_index] Update BWT... 13.04 sec\n",
      "[bwa_index] Pack forward-only FASTA... 10.96 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 1080.52 sec\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: bwa index ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa\n",
      "[main] Real time: 3194.229 sec; CPU: 3149.775 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start index genome already exists. Skipping downloading it.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "mkdir -p ./data/experiment\n",
    "if [[ ! -f ./data/experiment/SRR3997504_1.fastq || ! -f ./data/experiment/SRR3997504_2.fastq ]]; then\n",
    "    echo \"One or both FASTQ files do not exist in the data folder. Running fasterq-dump...\"\n",
    "    fasterq-dump -O ./data/experiment SRR3997504\n",
    "else\n",
    "    echo \"Both FASTQ files already exist in the data folder. Skipping fasterq-dump.\"\n",
    "fi\n",
    "if [[ ! -f ./data/experiment/SRR3997504_2_val_2_fastqc.html ]]; then\n",
    "    \"No trimmed files by trim_galore found. Running trim_galore...\"\n",
    "    trim_galore -j $((THREADS/4)) --paired --fastqc -o ./data/experiment/ data/experiment/SRR3997504_1.fastq ./data/experiment/SRR3997504_2.fastq\n",
    "else\n",
    "    echo \"Trimmed files by trim_galore already exist. Skipping trim_galore.\"\n",
    "fi\n",
    "\n",
    "#https://zenodo.org/records/5146236/files/star.tar.gz?download=1\n",
    "\n",
    "\n",
    "mkdir -p ./data/genome\n",
    "if [[ ! -f ./data/genome/hg38.fa ]]; then\n",
    "    echo \"USCS genome 38 does not exist in the data folder. Downloading, unziping and indexing it...\"\n",
    "    wget -q --show-progress --progress=bar:force:noscroll -P ./data/genome/ http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\n",
    "    pigz -cd -p $THREADS ./data/genome/hg38.fa.gz > ./data/genome/hg38.fa\n",
    "    bwa index ./data/genome/hg38.fa\n",
    "    rm -rf ./data/genome/hg38.fa.gz\n",
    "else\n",
    "    echo \"USCS genome 38 found. Assuming it is bwa indexed. Skipping downlaoding, unzipping and indexing it.\"\n",
    "fi\n",
    "\n",
    "\n",
    "if [[ ! -f ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa ]]; then \n",
    "    echo \"Downloading and unzipping Ensembl genome...\"\n",
    "    wget -q --show-progress --progress=bar:force:noscroll -P ./data/ http://itmat.data-simulators.s3.amazonaws.com/BEERS2/CAMPAREE_RESOURCE_FILES/HomoSapiens_GRCh38_Ensemblv99__Resource_files.tar.gz\n",
    "    pigz -cd -p $THREADS ./data/HomoSapiens_GRCh38_Ensemblv99__Resource_files.tar.gz | tar xf - -C ./data/\n",
    "    rm -rf ./data/HomoSapiens_GRCh38_Ensemblv99__Resource_files.tar.gz\n",
    "else\n",
    "    echo \"Ensembl genome exists. Skipping downloading it.\"\n",
    "fi\n",
    "if [[ ! -f ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa.fai ]]; then \n",
    "    echo \"Indexing the Ensembl genome file...\"\n",
    "    bwa index ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa\n",
    "else\n",
    "    echo \"Ensebml genome file was indexed. Skipping indexing it.\"\n",
    "fi\n",
    "if [[ ! -f ./data/HomoSapiens_GRCh38_Ensemblv99/star_index.genome/Genome ]]; then \n",
    "    echo \"Downloading and unzipping star index genome...\"\n",
    "    wget -q --show-progress --progress=bar:force:noscroll -P ./data/ http://itmat.data-simulators.s3.amazonaws.com/BEERS2/CAMPAREE_RESOURCE_FILES/HomoSapiens_GRCh38_Ensemblv99__STAR_index.tar.gz\n",
    "    pigz -cd -p $THREADS ./data/HomoSapiens_GRCh38_Ensemblv99__STAR_index.tar.gz | tar xf - -C ./data/\n",
    "    rm -rf ./data/HomoSapiens_GRCh38_Ensemblv99__STAR_index.tar.gz\n",
    "else\n",
    "    echo \"Start index genome already exists. Skipping downloading it.\"\n",
    "fi \n",
    "\n",
    "\n",
    "if [[ ! -f ./data/experiment/SRR3997504.sam ]]; then\n",
    "    hisat2-build --quiet -p 6 ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs\n",
    "    hisat2 --quiet -x ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs -1 ./data/experiment/SRR3997504_1.fastq -2 ./data/experiment/SRR3997504_2.fastq -S ./data/experiment/SRR3997504.sam -p 6\n",
    "fi\n",
    "if [[ ! -f ./data/experiment/SRR3997504.bam ]]; then\n",
    "    samtools view -bS -o ./data/experiment/SRR3997504.bam ./data/experiment/SRR3997504.sam\n",
    "fi\n",
    "if [[ ! -f ./data/experiment/SRR3997504_sorted.bam ]]; then\n",
    "    samtools sort ./data/experiment/SRR3997504.bam > ./data/experiment/SRR3997504_sorted.bam\n",
    "fi\n",
    "if [[ ! -f ./data/experiment/SRR3997504_sorted.bam.bai ]]; then\n",
    "    samtools index ./data/experiment/SRR3997504_sorted.bam\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e04a6fd6",
   "metadata": {},
   "source": [
    "We then skip the step of running the generate_L1EM_fasta_and_index.sh by running the commands to adjust for proper paths and skip it if it is already done."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "9ce5ba00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "L1EM.400.fa already exists. Skipping its creation.\n",
      "L1EM.400.fa was already indexed. Skipping its indexing.\n",
      "L1EM.400.ens.fa already exists. Skipping its creation.\n",
      "L1EM.400.ens.fa was already indexed. Skipping its indexing.\n",
      "Creating L1EM.400.ens.noY.fa from L1EM.400.ens.noY.bed...\n",
      "Indexing the L1EM.400.ens.noY.fa...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 1.04 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=494063968, availableWord=46763976\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 75938048 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 141444032 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 199661936 characters processed.\n",
      "[BWTIncConstructFromPacked] 40 iterations done. 251402128 characters processed.\n",
      "[BWTIncConstructFromPacked] 50 iterations done. 297384928 characters processed.\n",
      "[BWTIncConstructFromPacked] 60 iterations done. 338250528 characters processed.\n",
      "[BWTIncConstructFromPacked] 70 iterations done. 374567952 characters processed.\n",
      "[BWTIncConstructFromPacked] 80 iterations done. 406842944 characters processed.\n",
      "[BWTIncConstructFromPacked] 90 iterations done. 435525024 characters processed.\n",
      "[BWTIncConstructFromPacked] 100 iterations done. 461013680 characters processed.\n",
      "[BWTIncConstructFromPacked] 110 iterations done. 483664032 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 116 iterations.\n",
      "[bwa_index] 132.16 seconds elapse.\n",
      "[bwa_index] Update BWT... 1.12 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.66 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 55.39 sec\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: bwa index ./L1EM/annotation/L1EM.400.ens.noY.fa\n",
      "[main] Real time: 191.889 sec; CPU: 190.369 sec\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "if [[ ! -f ./L1EM/annotation/L1EM.400.fa ]]; then \n",
    "    echo \"Creating L1EM.400.fa from L1EM.400.bed...\"\n",
    "    bedtools getfasta -s -name -fi ./data/genome/hg38.fa -bed ./L1EM/annotation/L1EM.400.bed > ./L1EM/annotation/L1EM.400.fa\n",
    "else\n",
    "    echo \"L1EM.400.fa already exists. Skipping its creation.\"\n",
    "fi\n",
    "if [[ ! -f ./L1EM/annotation/L1EM.400.fa.sa ]]; then \n",
    "    echo \"Indexing the L1EM.400.fa...\"\n",
    "    bwa index ./L1EM/annotation/L1EM.400.fa\n",
    "else\n",
    "    echo \"L1EM.400.fa was already indexed. Skipping its indexing.\"\n",
    "fi\n",
    "\n",
    "\n",
    "if [[ ! -f ./L1EM/annotation/L1EM.400.ens.bed ]]; then \n",
    "    sed -E 's/^(chr)([^[:space:]]+)/\\2/' ./L1EM/annotation/L1EM.400.bed > ./L1EM/annotation/L1EM.400.ens.bed\n",
    "fi\n",
    "if [[ ! -f ./L1EM/annotation/L1EM.400.ens.fa ]]; then \n",
    "    echo \"Creating L1EM.400.ens.fa from L1EM.400.ens.bed...\"\n",
    "    bedtools getfasta -s -name -fi ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa -bed ./L1EM/annotation/L1EM.400.ens.bed > ./L1EM/annotation/L1EM.400.ens.fa\n",
    "else\n",
    "    echo \"L1EM.400.ens.fa already exists. Skipping its creation.\"\n",
    "fi\n",
    "if [[ ! -f ./L1EM/annotation/L1EM.400.ens.fa.sa ]]; then \n",
    "    echo \"Indexing the L1EM.400.ens.fa...\"\n",
    "    bwa index ./L1EM/annotation/L1EM.400.ens.fa\n",
    "else\n",
    "    echo \"L1EM.400.ens.fa was already indexed. Skipping its indexing.\"\n",
    "fi\n",
    "\n",
    "\n",
    "if [[ ! -f ./L1EM/annotation/L1EM.400.ens.noY.bed ]]; then \n",
    "    awk '$1 != \"Y\" && $1 != \"chrY\"' ./L1EM/annotation/L1EM.400.ens.bed > ./L1EM/annotation/L1EM.400.ens.noY.bed\n",
    "fi\n",
    "if [[ ! -f ./L1EM/annotation/L1EM.400.ens.noY.fa ]]; then \n",
    "    echo \"Creating L1EM.400.ens.noY.fa from L1EM.400.ens.noY.bed...\"\n",
    "    bedtools getfasta -s -name -fi ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa -bed ./L1EM/annotation/L1EM.400.ens.noY.bed > ./L1EM/annotation/L1EM.400.ens.noY.fa\n",
    "else\n",
    "    echo \"L1EM.400.ens.noY.fa already exists. Skipping its creation.\"\n",
    "fi\n",
    "if [[ ! -f ./L1EM/annotation/L1EM.400.ens.noY.fa.sa ]]; then \n",
    "    echo \"Indexing the L1EM.400.ens.noY.fa...\"\n",
    "    bwa index ./L1EM/annotation/L1EM.400.ens.noY.fa\n",
    "else\n",
    "    echo \"L1EM.400.ens.noY.fa was already indexed. Skipping its indexing.\"\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dfa28ed",
   "metadata": {},
   "source": [
    "We then have to generate the appropriate annotation files. However, as pygtftk is not updated and has conflicting requirements with the rest of the project we are going to be providing the resulting file from runnning the command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63a5542b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!gtftk bed_to_gtf -i ./L1EM/annotation/L1EM.400.bed -o ./annotations/original.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a9ec1ce",
   "metadata": {},
   "source": [
    "We then continue to do various edits that are needed for the various runs that we are going to do. We also have to remove the colons (:) as Camparee is not compatible with them althouth they are not against the gtf standard."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "f2968594",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Make the colons into underscores\n",
    "if [[ ! -f ./annotations/nocolon.gtf ]]; then \n",
    "    sed -E 's/(\\.chr([1-9]|1[0-9]|2[0-2]|x|y|mt|un)):/\\1_/gI' ./annotations/original.gtf > ./annotations/nocolon.gtf\n",
    "fi\n",
    "# Prepare for the camparee format\n",
    "if [[ ! -f ./annotations/exons.gtf ]]; then \n",
    "    sed 's/\\ttranscript\\t/\\texon\\t/g' ./annotations/nocolon.gtf > ./annotations/exons.gtf\n",
    "fi\n",
    "mkdir -p ./annotations/camparee\n",
    "# if [[ ! -f ./data/HomoSapiens_GRCh38_Ensemblv99/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz ]]; then \n",
    "#     wget -q --show-progress --progress=bar:force:noscroll -P ./data/HomoSapiens_GRCh38_Ensemblv99/ ftp://ftp.ensembl.org/pub/release-99/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
    "# fi\n",
    "# if [[ ! -f ./data/HomoSapiens_GRCh38_Ensemblv99/Homo_sapiens.GRCh38.99.gtf.gz ]]; then \n",
    "#     wget -q --show-progress --progress=bar:force:noscroll -P ./data/HomoSapiens_GRCh38_Ensemblv99/ ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz\n",
    "# fi\n",
    "if [[ ! -f ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.txt ]]; then \n",
    "    $CONDA_PREFIX/lib/python3.11/site-packages/camparee/bin/format_reference_files_for_camparee.py -n HomoSapiens_hg38 -o ./annotations/camparee/ -g ./data/genome/hg38.fa -a ./annotations/exons.gtf -p\n",
    "fi\n",
    "if [[ ! -f ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.chrNamesfix.txt ]]; then \n",
    "    # sed -E 's/^(chr)([1-9]|1[0-9]|2[0-2]|x|y|mt|un)(\\t)/\\3_/' ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.txt > ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.chrNamesfix.txt\n",
    "    sed -E 's/^(chr)([^[:space:]]+)/\\2/' ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.txt > ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.chrNamesfix.txt\n",
    "fi\n",
    "if [[ ! -f ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.sorted.txt ]]; then \n",
    "    sort -k1,1 -k4,4n -k5,5n ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.chrNamesfix.txt > ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.sorted.txt\n",
    "fi\n",
    "if [[ ! -f ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.ashes.txt ]]; then \n",
    "    grep -E '^#|L[^ \\t]*\\.1\\.' ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.sorted.txt > ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.ashes.txt\n",
    "    python ./src/annotation400.py ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.ashes.txt ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.ashes.400.txt\n",
    "fi\n",
    "\n",
    "cp ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.ashes.400.txt ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_hg38.annotation.ashes.txt\n",
    "cp ./annotations/camparee/HomoSapiens_hg38/HomoSapiens_hg38.annotation.sorted.txt ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_hg38.annotation.mixed.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d8e582",
   "metadata": {},
   "source": [
    "We will try to run the flux-simulator now.\n",
    "\n",
    "To do so we firstly need to create a temporary file for it, split our fasta file into different transcript files and run it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "819904c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p /tmp/fluxsim\n",
    "flux-simulator -p ./data/flux/flux.par -x -l -s\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "860ba964",
   "metadata": {},
   "source": [
    "Then we want to do 2 different runs of camparee (following the paper) to simualte the runs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "c522f6a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running CAMPAREE using the serial job scheduler.\n",
      "With the following default scheduler parameters:\n",
      " \t-default_memory_in_mb : 26000\n",
      "\t-default_submission_args : None\n",
      "\t-default_num_processors : 10\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "And a maximum job resubmission limit of 3.\n",
      "Execution of the Expression Pipeline Started...\n",
      "Running jobs:0 | Pending jobs:5 | Resub jobs:0 | Completed jobs:0\n",
      "--Check 5 pending jobs for satisfied dependencies:\n",
      "\tSubmitting GenomeAlignmentStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting GenomeAlignmentStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:4 | Resub jobs:0 | Completed jobs:1\n",
      "--Check 4 pending jobs for satisfied dependencies:\n",
      "\tSubmitting GenomeBamIndexStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting GenomeBamIndexStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:3 | Resub jobs:0 | Completed jobs:2\n",
      "--Check 3 pending jobs for satisfied dependencies:\n",
      "\tSubmitting VariantsFinderStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting VariantsFinderStep command to serial for sample SRR3997504.\n",
      "\tSubmitting IntronQuantificationStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting IntronQuantificationStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:1 | Resub jobs:0 | Completed jobs:4\n",
      "--Check 1 pending jobs for satisfied dependencies:\n",
      "\tSubmitting VariantsCompilationStep command to serial.\n",
      "\tFinished submitting VariantsCompilationStep command to serial.\n",
      "Running jobs:0 | Pending jobs:0 | Resub jobs:0 | Completed jobs:5\n",
      "Processing sample1 (SRR3997504)...\n",
      "Running jobs:0 | Pending jobs:16 | Resub jobs:0 | Completed jobs:5\n",
      "--Check 16 pending jobs for satisfied dependencies:\n",
      "\tSubmitting GenomeBuilderStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting GenomeBuilderStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:15 | Resub jobs:0 | Completed jobs:6\n",
      "--Check 15 pending jobs for satisfied dependencies:\n",
      "\tSubmitting UpdateAnnotationForGenomeStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting UpdateAnnotationForGenomeStep command to serial for sample SRR3997504.\n",
      "\tSubmitting UpdateAnnotationForGenomeStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting UpdateAnnotationForGenomeStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:13 | Resub jobs:0 | Completed jobs:8\n",
      "--Check 13 pending jobs for satisfied dependencies:\n",
      "\tSubmitting TranscriptomeFastaPreparationStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting TranscriptomeFastaPreparationStep command to serial for sample SRR3997504.\n",
      "\tSubmitting TranscriptomeFastaPreparationStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting TranscriptomeFastaPreparationStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:11 | Resub jobs:0 | Completed jobs:10\n",
      "--Check 11 pending jobs for satisfied dependencies:\n",
      "\tSubmitting KallistoIndexStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting KallistoIndexStep command to serial for sample SRR3997504.\n",
      "\tSubmitting Bowtie2IndexStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting Bowtie2IndexStep command to serial for sample SRR3997504.\n",
      "\tSubmitting KallistoIndexStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting KallistoIndexStep command to serial for sample SRR3997504.\n",
      "\tSubmitting Bowtie2IndexStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting Bowtie2IndexStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:7 | Resub jobs:0 | Completed jobs:14\n",
      "--Check 7 pending jobs for satisfied dependencies:\n",
      "\tSubmitting KallistoQuantStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting KallistoQuantStep command to serial for sample SRR3997504.\n",
      "\tSubmitting Bowtie2AlignStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting Bowtie2AlignStep command to serial for sample SRR3997504.\n",
      "\tSubmitting KallistoQuantStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting KallistoQuantStep command to serial for sample SRR3997504.\n",
      "\tSubmitting Bowtie2AlignStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting Bowtie2AlignStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:3 | Resub jobs:0 | Completed jobs:18\n",
      "--Check 3 pending jobs for satisfied dependencies:\n",
      "\tSubmitting TranscriptGeneQuantificationStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting TranscriptGeneQuantificationStep command to serial for sample SRR3997504.\n",
      "\tSubmitting AllelicImbalanceQuantificationStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting AllelicImbalanceQuantificationStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:1 | Resub jobs:0 | Completed jobs:20\n",
      "--Check 1 pending jobs for satisfied dependencies:\n",
      "\tSubmitting MoleculeMakerStep command to serial for sample SRR3997504.\n",
      "\tFinished submitting MoleculeMakerStep command to serial for sample SRR3997504.\n",
      "Running jobs:0 | Pending jobs:0 | Resub jobs:0 | Completed jobs:21\n",
      "Execution of the Expression Pipeline Ended\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "if [[ ! -f ./output/run_1/CAMPAREE/data/sample1/molecule_file.txt ]]; then\n",
    "    rm -rf ./output/run_1\n",
    "    camparee -c ./configs/camparee.ashes.yaml -r1\n",
    "fi\n",
    "# if [[ ! -f ./output/run_2/CAMPAREE/data/sample1/molecule_file.txt ]]; then \n",
    "#     rm -rf ./output/run_2\n",
    "#     camparee -c ./configs/camparee.mixed.yaml -r2\n",
    "# fi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0523215c",
   "metadata": {},
   "source": [
    "We then use beer2 to account for the errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "42e8b574",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mCreating specified working directory ./output/beers2ashes.\u001b[0m\n",
      "\u001b[33mBuilding DAG of jobs...\u001b[0m\n",
      "\u001b[33mUsing shell: /usr/bin/bash\u001b[0m\n",
      "\u001b[33mProvided cores: 4\u001b[0m\n",
      "\u001b[33mRules claiming more threads will be scaled down.\u001b[0m\n",
      "\u001b[33mJob stats:\n",
      "job                                          count    min threads    max threads\n",
      "-----------------------------------------  -------  -------------  -------------\n",
      "all                                              1              1              1\n",
      "create_cluster_packet                            1              1              1\n",
      "create_sequencer_outputs_sam_or_bam              1              1              1\n",
      "run_library_prep_packet_from_distribution        1              1              1\n",
      "sequence_cluster_packet                          1              1              1\n",
      "total                                            5              1              1\n",
      "\u001b[0m\n",
      "\u001b[33mSelect jobs to execute...\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "\u001b[32m[Thu Jul 10 14:02:45 2025]\u001b[0m\n",
      "\u001b[32mrule run_library_prep_packet_from_distribution:\n",
      "    input: ../../output/run_1/CAMPAREE/data/sample1/custom_genome_1.fa, ../../output/run_1/CAMPAREE/data/sample1/custom_genome_2.fa, ../../output/run_1/CAMPAREE/data/sample1/custom_genome_indels_1.txt, ../../output/run_1/CAMPAREE/data/sample1/custom_genome_indels_2.txt, ../../output/run_1/CAMPAREE/data/sample1/updated_annotation_1.txt, ../../output/run_1/CAMPAREE/data/sample1/updated_annotation_2.txt, ../../output/run_1/CAMPAREE/data/sample1/intron_quantifications.txt, ../../output/run_1/CAMPAREE/data/sample1/gene_quantifications.txt, ../../output/run_1/CAMPAREE/data/sample1/isoform_psi_value_quantifications.txt, ../../output/run_1/CAMPAREE/data/sample1/allelic_imbalance_quantifications.txt\n",
      "    output: library_prep_pipeline/sample1/from_distribution/library_prep_pipeline_result_molecule_pkt0.txt, library_prep_pipeline/sample1/from_distribution/library_prep_pipeline_result_molecule_pkt0.quant_file, library_prep_pipeline/sample1/from_distribution/input_quants/library_prep_pipeline_molecule_pkt0.quant_file, library_prep_pipeline/sample1/logs/PolyAStep/molecule_pkt0.log, library_prep_pipeline/sample1/logs/FragmentStep/molecule_pkt0.log, library_prep_pipeline/sample1/logs/FirstStrandSynthesisStep/molecule_pkt0.log, library_prep_pipeline/sample1/logs/SecondStrandSynthesisStep/molecule_pkt0.log, library_prep_pipeline/sample1/logs/SizingStep/molecule_pkt0.log, library_prep_pipeline/sample1/logs/AdapterLigationStep/molecule_pkt0.log, library_prep_pipeline/sample1/logs/PCRAmplificationStep/molecule_pkt0.log\n",
      "    jobid: 4\n",
      "    reason: Missing output files: library_prep_pipeline/sample1/from_distribution/library_prep_pipeline_result_molecule_pkt0.txt\n",
      "    wildcards: sample=1, packet_num=0\n",
      "    resources: tmpdir=/tmp, mem_mb=12000, mem_mib=11445\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "Initializing with seed [100, '1', 0, 1]\n",
      "/home/bio/git/l1emPipeline/output/beers2ashes/library_prep_pipeline/sample1/logs/sample1\n",
      "Generating molecules for sample1.\n",
      "Loading gene, intron, transcript PSI, and allelic imbalance distributions.\n",
      "Loading annotations, transcriptome sequences, and genome sequences from both parental genomes.\n",
      "Loading indel information from both parental genomes.\n",
      "Generating molecules and saving/exporting the results.\n",
      "Molecule maker output type 'generator'\n",
      "Generating 1 packets\n",
      "Sample has 30000 molecules\n",
      "Percent of the original ids that are still represented: 0.00%\n",
      "Counts of molecules in size ranges:\n",
      " <=100: 0\n",
      " <=500: 0\n",
      " <=1000: 306\n",
      ">1000: 29694\n",
      "Poly A selection step instantiated\n",
      "First Strand cDNA Synthesis Step instantiated.\n",
      "Second Strand cDNA Synthesis Step instantiated.\n",
      "Sizing step instantiated\n",
      "Adapter Ligation Step instantiated\n",
      "Opening library_prep_pipeline/sample1/logs/PolyAStep/molecule_pkt0.log\n",
      "Poly A selection step starting\n",
      "Poly A selection step complete\n",
      "Step took 2.38 seconds\n",
      "Sample has 30000 molecules\n",
      "Percent of the original ids that are still represented: 0.00%\n",
      "Counts of molecules in size ranges:\n",
      " <=100: 0\n",
      " <=500: 0\n",
      " <=1000: 306\n",
      ">1000: 29694\n",
      "Opening library_prep_pipeline/sample1/logs/FragmentStep/molecule_pkt0.log\n",
      "Fragment Step acting on sample\n",
      "Step took 9.66 seconds\n",
      "Sample has 789766 molecules\n",
      "Percent of the original ids that are still represented: 0.00%\n",
      "Counts of molecules in size ranges:\n",
      " <=100: 267789\n",
      " <=500: 456065\n",
      " <=1000: 61166\n",
      ">1000: 4746\n",
      "Opening library_prep_pipeline/sample1/logs/FirstStrandSynthesisStep/molecule_pkt0.log\n",
      "First strand cDNA synthesis step starting...\n",
      "First strand cDNA synthesis step complete.\n",
      "Step took 1.18e+02 seconds\n",
      "Sample has 789766 molecules\n",
      "Percent of the original ids that are still represented: 0.00%\n",
      "Counts of molecules in size ranges:\n",
      " <=100: 314513\n",
      " <=500: 415530\n",
      " <=1000: 55428\n",
      ">1000: 4295\n",
      "Opening library_prep_pipeline/sample1/logs/SecondStrandSynthesisStep/molecule_pkt0.log\n",
      "Second strand cDNA synthesis step starting...\n",
      "Second strand cDNA synthesis step complete.\n",
      "Step took 1.1e+02 seconds\n",
      "Sample has 789766 molecules\n",
      "Percent of the original ids that are still represented: 0.00%\n",
      "Counts of molecules in size ranges:\n",
      " <=100: 357724\n",
      " <=500: 378055\n",
      " <=1000: 50102\n",
      ">1000: 3885\n",
      "Opening library_prep_pipeline/sample1/logs/SizingStep/molecule_pkt0.log\n",
      "Sizing step starting\n",
      "Sizing step complete\n",
      "Step took 0.359 seconds\n",
      "Sample has 217297 molecules\n",
      "Percent of the original ids that are still represented: 0.00%\n",
      "Counts of molecules in size ranges:\n",
      " <=100: 0\n",
      " <=500: 217297\n",
      " <=1000: 0\n",
      ">1000: 0\n",
      "Opening library_prep_pipeline/sample1/logs/AdapterLigationStep/molecule_pkt0.log\n",
      "Adapter Ligation Step starting\n",
      "Step took 1.4 seconds\n",
      "Sample has 217297 molecules\n",
      "Percent of the original ids that are still represented: 0.00%\n",
      "Counts of molecules in size ranges:\n",
      " <=100: 0\n",
      " <=500: 213826\n",
      " <=1000: 3471\n",
      ">1000: 0\n",
      "Opening library_prep_pipeline/sample1/logs/PCRAmplificationStep/molecule_pkt0.log\n",
      "PCR Amplication step starting\n",
      "Starting cycle 1\n",
      "Starting cycle 2\n",
      "Starting cycle 3\n",
      "Starting cycle 4\n",
      "Starting cycle 5\n",
      "Starting cycle 6\n",
      "Starting cycle 7\n",
      "Starting cycle 8\n",
      "Starting cycle 9\n",
      "Starting cycle 10\n",
      "Step took 6.92 seconds\n",
      "Sample has 47900 molecules\n",
      "Percent of the original ids that are still represented: 0.00%\n",
      "Counts of molecules in size ranges:\n",
      " <=100: 0\n",
      " <=500: 47484\n",
      " <=1000: 416\n",
      ">1000: 0\n",
      "Finished library_prep_pipeline in 250.7 seconds\n",
      "Output final sample to library_prep_pipeline/sample1/from_distribution/library_prep_pipeline_result_molecule_pkt0.txt\n",
      "Output final sample quantification to library_prep_pipeline/sample1/from_distribution/library_prep_pipeline_result_molecule_pkt0.quant_file\n",
      "\u001b[32m[Thu Jul 10 14:08:32 2025]\u001b[0m\n",
      "\u001b[32mFinished job 4.\u001b[0m\n",
      "\u001b[32m1 of 5 steps (20%) done\u001b[0m\n",
      "\u001b[33mSelect jobs to execute...\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "\u001b[32m[Thu Jul 10 14:08:32 2025]\u001b[0m\n",
      "\u001b[32mrule create_cluster_packet:\n",
      "    input: library_prep_pipeline/sample1/from_distribution/library_prep_pipeline_result_molecule_pkt0.txt\n",
      "    output: sequence_pipeline/sample1/input_cluster_packets/cluster_packet_start_pkt0.gzip\n",
      "    jobid: 3\n",
      "    reason: Missing output files: sequence_pipeline/sample1/input_cluster_packets/cluster_packet_start_pkt0.gzip; Input files updated by another job: library_prep_pipeline/sample1/from_distribution/library_prep_pipeline_result_molecule_pkt0.txt\n",
      "    wildcards: sample=1, packet_num=0\n",
      "    resources: tmpdir=/tmp\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "Loading library_prep_pipeline/sample1/from_distribution/library_prep_pipeline_result_molecule_pkt0.txt to flowcell\n",
      "Assigned flowcell coordinates to 47900 clusters.\n",
      "Output cluster packet to sequence_pipeline/sample1/input_cluster_packets/cluster_packet_start_pkt0.gzip\n",
      "Intermediate loaded - process RAM at 0.165688 GB\n",
      "\u001b[32m[Thu Jul 10 14:08:38 2025]\u001b[0m\n",
      "\u001b[32mFinished job 3.\u001b[0m\n",
      "\u001b[32m2 of 5 steps (40%) done\u001b[0m\n",
      "\u001b[33mSelect jobs to execute...\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "\u001b[32m[Thu Jul 10 14:08:38 2025]\u001b[0m\n",
      "\u001b[32mrule sequence_cluster_packet:\n",
      "    input: sequence_pipeline/sample1/input_cluster_packets/cluster_packet_start_pkt0.gzip\n",
      "    output: sequence_pipeline/sample1/output_cluster_packets/sequence_cluster_packet0.gzip, sequence_pipeline/sample1/logs/BridgeAmplificationStep/sequence_pipeline_cluster_pkt0.log, sequence_pipeline/sample1/logs/SequenceBySynthesisStep/sequence_pipeline_cluster_pkt0.log\n",
      "    jobid: 2\n",
      "    reason: Missing output files: sequence_pipeline/sample1/output_cluster_packets/sequence_cluster_packet0.gzip; Input files updated by another job: sequence_pipeline/sample1/input_cluster_packets/cluster_packet_start_pkt0.gzip\n",
      "    wildcards: sample=1, packet_num=0\n",
      "    resources: tmpdir=/tmp, mem_mb=12000, mem_mib=11445\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "Initializing with seed [100, '1', 0, 2]\n",
      "Bridge Amplification Step instantiated\n",
      "Sequence By Synthesis Step instantiated\n",
      "Execution of the sequence_pipeline Started...\n",
      "Molecules per cluster after amplification is 1024.\n",
      "Total molecules over all clusters is 49049600\n",
      "Starting Sequence By Synthesis Step\n",
      "Finished sequence pipeline in 84.5 seconds\n",
      "Output final sample to sequence_pipeline/sample1/output_cluster_packets/sequence_cluster_packet0.gzip\n",
      "\u001b[32m[Thu Jul 10 14:11:04 2025]\u001b[0m\n",
      "\u001b[32mFinished job 2.\u001b[0m\n",
      "\u001b[32m3 of 5 steps (60%) done\u001b[0m\n",
      "\u001b[33mSelect jobs to execute...\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "\u001b[32m[Thu Jul 10 14:11:04 2025]\u001b[0m\n",
      "\u001b[32mrule create_sequencer_outputs_sam_or_bam:\n",
      "    input: sequence_pipeline/sample1/output_cluster_packets/sequence_cluster_packet0.gzip, ../../data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa\n",
      "    output: results/S1_L1.bam, results/S1_L2.bam, results/S1_unidentifiedL1.bam, results/S1_unidentifiedL2.bam\n",
      "    jobid: 1\n",
      "    reason: Missing output files: results/S1_L1.bam, results/S1_L2.bam; Input files updated by another job: sequence_pipeline/sample1/output_cluster_packets/sequence_cluster_packet0.gzip\n",
      "    wildcards: sample=1, sequencer_output_filetype=bam\n",
      "    resources: tmpdir=/tmp, mem_mb=10000, mem_mib=9537\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "Loading from sequence_pipeline/sample1/output_cluster_packets/sequence_cluster_packet0.gzip\n",
      "Writing out demultiplexed sam files to:\n",
      "results/S1_L1.bam\n",
      "results/S1_L2.bam\n",
      "\u001b[32m[Thu Jul 10 14:11:46 2025]\u001b[0m\n",
      "\u001b[32mFinished job 1.\u001b[0m\n",
      "\u001b[32m4 of 5 steps (80%) done\u001b[0m\n",
      "\u001b[33mSelect jobs to execute...\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "\u001b[32m[Thu Jul 10 14:11:46 2025]\u001b[0m\n",
      "\u001b[32mlocalrule all:\n",
      "    input: results/S1_L1.bam, results/S1_L2.bam\n",
      "    jobid: 0\n",
      "    reason: Input files updated by another job: results/S1_L1.bam, results/S1_L2.bam\n",
      "    resources: tmpdir=/tmp\u001b[0m\n",
      "\u001b[32m\u001b[0m\n",
      "\u001b[32m[Thu Jul 10 14:11:46 2025]\u001b[0m\n",
      "\u001b[32mFinished job 0.\u001b[0m\n",
      "\u001b[32m5 of 5 steps (100%) done\u001b[0m\n",
      "\u001b[33mComplete log: .snakemake/log/2025-07-10T140244.093327.snakemake.log\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "# %%bash\n",
    "# %conda install -y numpy=1.25.2\n",
    "# %pip install git+https://github.com/itmat/BEERS2\n",
    "# if [[ ! -f ./output/beers2ashes/results/S1_L2.bam ]]; then \n",
    "!beers --configfile ./configs/beers2.ashes.yaml --jobs 4 --directory ./output/beers2ashes\n",
    "# fi\n",
    "# if [[ ! -f ./output/beers2mixed/results/S1_L2.bam ]]; then \n",
    "# !beers --configfile ./configs/beers2.mixed.yaml --jobs 4 --directory ./output/beers2mixed\n",
    "# fi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2e49a0d",
   "metadata": {},
   "source": [
    "We then take the output of the beers and we merge the two output files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6e78a0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexing merged results from beers (ashes)...\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "if [[ ! -f ./output/beers2ashes/results/S1_L1.sorted.bam ]]; then \n",
    "    samtools sort ./output/beers2ashes/results/S1_L1.bam > ./output/beers2ashes/results/S1_L1.sorted.bam\n",
    "fi\n",
    "if [[ ! -f ./output/beers2ashes/results/S1_L2.sorted.bam ]]; then \n",
    "    samtools sort ./output/beers2ashes/results/S1_L2.bam > ./output/beers2ashes/results/S1_L2.sorted.bam\n",
    "fi\n",
    "if [[ ! -f ./output/beers2ashes/results/merged.bam ]]; then \n",
    "    samtools merge -o ./output/beers2ashes/results/merged.bam ./output/beers2ashes/results/S1_L1.sorted.bam ./output/beers2ashes/results/S1_L2.sorted.bam\n",
    "fi\n",
    "if [[ ! -f ./output/beers2ashes/results/merged.sorted.bam ]]; then \n",
    "    samtools sort ./output/beers2ashes/results/merged.bam > ./output/beers2ashes/results/merged.sorted.bam\n",
    "fi\n",
    "if [[ ! -f ./output/beers2ashes/results/merged.sorted.bam.bai ]]; then \n",
    "    echo \"Indexing merged results from beers (ashes)...\"\n",
    "    samtools index ./output/beers2ashes/results/merged.sorted.bam\n",
    "fi\n",
    "\n",
    "# if [[ ! -f ./output/beers2mixed/results/S1_L1.sorted.bam ]]; then \n",
    "#     samtools sort ./output/beers2mixed/results/S1_L1.bam > ./output/beers2mixed/results/S1_L1.sorted.bam\n",
    "# fi\n",
    "# if [[ ! -f ./output/beers2mixed/results/S1_L2.sorted.bam ]]; then \n",
    "#     samtools sort ./output/beers2mixed/results/S1_L2.bam > ./output/beers2mixed/results/S1_L2.sorted.bam\n",
    "# fi\n",
    "# if [[ ! -f ./output/beers2mixed/results/merged.bam ]]; then \n",
    "#     samtools merge -o ./output/beers2mixed/results/merged.bam ./output/beers2mixed/results/S1_L1.sorted.bam ./output/beers2mixed/results/S1_L2.sorted.bam\n",
    "# fi\n",
    "# if [[ ! -f ./output/beers2mixed/results/merged.sorted.bam ]]; then \n",
    "#     samtools sort ./output/beers2mixed/results/merged.bam > ./output/beers2mixed/results/merged.sorted.bam\n",
    "# fi\n",
    "# if [[ ! -f ./output/beers2mixed/results/merged.sorted.bam.bai ]]; then \n",
    "#     echo \"Indexing merged results from beers (mixed)...\"\n",
    "#     samtools index ./output/beers2mixed/results/merged.sorted.bam\n",
    "# fi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a18ad1",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73818224",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting git+https://github.com/itmat/BEERS2\n",
      "  Cloning https://github.com/itmat/BEERS2 to /tmp/pip-req-build-mk7vsj4b\n",
      "  Running command git clone --filter=blob:none --quiet https://github.com/itmat/BEERS2 /tmp/pip-req-build-mk7vsj4b\n",
      "  Resolved https://github.com/itmat/BEERS2 to commit 0fdd2f7db38c1c6858aab415dc80a1e685afe788\n",
      "  Running command git submodule update --init --recursive -q\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: torch in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (2.1.2)\n",
      "Collecting torch\n",
      "  Using cached torch-2.7.1-cp311-cp311-manylinux_2_28_x86_64.whl.metadata (29 kB)\n",
      "Collecting jax\n",
      "  Downloading jax-0.6.2-py3-none-any.whl.metadata (13 kB)\n",
      "Collecting numpyro\n",
      "  Downloading numpyro-0.18.0-py3-none-any.whl.metadata (37 kB)\n",
      "Collecting beers_utils@ git+https://github.com/itmat/BEERS_UTILS@52e1a00057d2939651ab587a4ad1a6873dc8d49e (from BEERS2==2.0)\n",
      "  Using cached beers_utils-0.2.0-py3-none-any.whl\n",
      "Collecting camparee@ git+https://github.com/itmat/CAMPAREE@1bb79db71ea8d74495334f1af74025624a4668c7 (from BEERS2==2.0)\n",
      "  Using cached camparee-0.4.2-py3-none-any.whl\n",
      "Requirement already satisfied: appdirs==1.4.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.4.4)\n",
      "Requirement already satisfied: attrs==22.2.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (22.2.0)\n",
      "Requirement already satisfied: boto3==1.26.85 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.26.85)\n",
      "Requirement already satisfied: botocore==1.29.85 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.29.85)\n",
      "Requirement already satisfied: certifi==2022.12.7 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2022.12.7)\n",
      "Requirement already satisfied: charset-normalizer==3.1.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.1.0)\n",
      "Requirement already satisfied: cloudpathlib==0.13.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.13.0)\n",
      "Requirement already satisfied: ConfigArgParse==1.5.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.5.3)\n",
      "Requirement already satisfied: connection-pool==0.0.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.0.3)\n",
      "Requirement already satisfied: contourpy==1.0.7 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.0.7)\n",
      "Requirement already satisfied: cycler==0.11.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.11.0)\n",
      "Requirement already satisfied: Cython==0.29.33 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.29.33)\n",
      "Requirement already satisfied: datrie==0.8.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.8.2)\n",
      "Requirement already satisfied: docutils==0.19 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.19)\n",
      "Requirement already satisfied: dpath==2.1.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.1.4)\n",
      "Requirement already satisfied: fastjsonschema==2.16.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.16.3)\n",
      "Requirement already satisfied: fonttools==4.39.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.39.0)\n",
      "Requirement already satisfied: gitdb==4.0.10 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.0.10)\n",
      "Requirement already satisfied: GitPython==3.1.31 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.1.31)\n",
      "Requirement already satisfied: humanfriendly==10.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (10.0)\n",
      "Requirement already satisfied: idna==3.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.4)\n",
      "Requirement already satisfied: Jinja2==3.1.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.1.2)\n",
      "Requirement already satisfied: jmespath==1.0.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.0.1)\n",
      "Requirement already satisfied: jsonschema==4.17.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.17.3)\n",
      "Requirement already satisfied: jupyter_core==5.2.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.2.0)\n",
      "Requirement already satisfied: kiwisolver==1.4.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.4.4)\n",
      "Requirement already satisfied: MarkupSafe==2.1.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.1.2)\n",
      "Requirement already satisfied: matplotlib==3.7.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.7.1)\n",
      "Requirement already satisfied: nbformat==5.7.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.7.3)\n",
      "Requirement already satisfied: numpy==1.24.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.24.2)\n",
      "Requirement already satisfied: packaging==23.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (23.0)\n",
      "Requirement already satisfied: pandas==1.5.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.5.3)\n",
      "Requirement already satisfied: Pillow==9.4.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (9.4.0)\n",
      "Requirement already satisfied: plac==1.3.5 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.3.5)\n",
      "Requirement already satisfied: platformdirs==3.1.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.1.0)\n",
      "Requirement already satisfied: prettytable==3.6.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.6.0)\n",
      "Requirement already satisfied: psutil==5.9.4 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.9.4)\n",
      "Requirement already satisfied: PuLP==2.7.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.7.0)\n",
      "Requirement already satisfied: pydantic==1.10.5 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.10.5)\n",
      "Requirement already satisfied: pyparsing==3.0.9 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (3.0.9)\n",
      "Requirement already satisfied: pyrsistent==0.19.3 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.19.3)\n",
      "Requirement already satisfied: pysam==0.20.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.20.0)\n",
      "Requirement already satisfied: python-dateutil==2.8.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.8.2)\n",
      "Requirement already satisfied: pytz==2022.7.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2022.7.1)\n",
      "Requirement already satisfied: PyYAML==6.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (6.0)\n",
      "Requirement already satisfied: requests==2.28.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.28.2)\n",
      "Requirement already satisfied: reretry==0.11.8 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.11.8)\n",
      "Requirement already satisfied: roman==4.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.0)\n",
      "Requirement already satisfied: s3transfer==0.6.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.6.0)\n",
      "Requirement already satisfied: scipy==1.10.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.10.1)\n",
      "Requirement already satisfied: six==1.16.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.16.0)\n",
      "Requirement already satisfied: smart-open==6.3.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (6.3.0)\n",
      "Requirement already satisfied: smmap==5.0.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.0.0)\n",
      "Requirement already satisfied: snakemake==7.24.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (7.24.0)\n",
      "Requirement already satisfied: stopit==1.1.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.1.2)\n",
      "Requirement already satisfied: tabulate==0.9.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.9.0)\n",
      "Requirement already satisfied: termcolor==2.2.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (2.2.0)\n",
      "Requirement already satisfied: throttler==1.2.2 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.2.2)\n",
      "Requirement already satisfied: toposort==1.10 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.10)\n",
      "Requirement already satisfied: traitlets==5.9.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (5.9.0)\n",
      "Requirement already satisfied: typing_extensions==4.5.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (4.5.0)\n",
      "Requirement already satisfied: urllib3==1.26.14 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.26.14)\n",
      "Requirement already satisfied: wcwidth==0.2.6 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (0.2.6)\n",
      "Requirement already satisfied: wrapt==1.15.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.15.0)\n",
      "Requirement already satisfied: yte==1.5.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from BEERS2==2.0) (1.5.1)\n",
      "Requirement already satisfied: filelock in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (3.18.0)\n",
      "INFO: pip is looking at multiple versions of torch to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting torch\n",
      "  Using cached torch-2.7.0-cp311-cp311-manylinux_2_28_x86_64.whl.metadata (29 kB)\n",
      "  Using cached torch-2.6.0-cp311-cp311-manylinux1_x86_64.whl.metadata (28 kB)\n",
      "  Using cached torch-2.5.1-cp311-cp311-manylinux1_x86_64.whl.metadata (28 kB)\n",
      "  Using cached torch-2.5.0-cp311-cp311-manylinux1_x86_64.whl.metadata (28 kB)\n",
      "  Using cached torch-2.4.1-cp311-cp311-manylinux1_x86_64.whl.metadata (26 kB)\n",
      "  Using cached torch-2.4.0-cp311-cp311-manylinux1_x86_64.whl.metadata (26 kB)\n",
      "  Using cached torch-2.3.1-cp311-cp311-manylinux1_x86_64.whl.metadata (26 kB)\n",
      "INFO: pip is still looking at multiple versions of torch to determine which version is compatible with other requirements. This could take a while.\n",
      "  Using cached torch-2.3.0-cp311-cp311-manylinux1_x86_64.whl.metadata (26 kB)\n",
      "  Using cached torch-2.2.2-cp311-cp311-manylinux1_x86_64.whl.metadata (25 kB)\n",
      "  Using cached torch-2.2.1-cp311-cp311-manylinux1_x86_64.whl.metadata (26 kB)\n",
      "  Using cached torch-2.2.0-cp311-cp311-manylinux1_x86_64.whl.metadata (25 kB)\n",
      "Requirement already satisfied: sympy in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (1.14.0)\n",
      "Requirement already satisfied: networkx in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (3.5)\n",
      "Requirement already satisfied: fsspec in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (2025.5.1)\n",
      "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (8.9.2.26)\n",
      "Requirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (12.1.3.1)\n",
      "Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (11.0.2.54)\n",
      "Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (10.3.2.106)\n",
      "Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (11.4.5.107)\n",
      "Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (12.1.0.106)\n",
      "Requirement already satisfied: nvidia-nccl-cu12==2.18.1 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (2.18.1)\n",
      "Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (12.1.105)\n",
      "Requirement already satisfied: triton==2.1.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from torch) (2.1.0)\n",
      "Requirement already satisfied: nvidia-nvjitlink-cu12 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from nvidia-cusolver-cu12==11.4.5.107->torch) (12.6.85)\n",
      "Collecting jaxlib<=0.6.2,>=0.6.2 (from jax)\n",
      "  Downloading jaxlib-0.6.2-cp311-cp311-manylinux2014_x86_64.whl.metadata (1.3 kB)\n",
      "Collecting ml_dtypes>=0.5.0 (from jax)\n",
      "  Downloading ml_dtypes-0.5.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (21 kB)\n",
      "INFO: pip is looking at multiple versions of jax to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting jax\n",
      "  Downloading jax-0.6.1-py3-none-any.whl.metadata (13 kB)\n",
      "Collecting jaxlib<=0.6.1,>=0.6.1 (from jax)\n",
      "  Downloading jaxlib-0.6.1-cp311-cp311-manylinux2014_x86_64.whl.metadata (1.2 kB)\n",
      "Collecting jax\n",
      "  Downloading jax-0.6.0-py3-none-any.whl.metadata (22 kB)\n",
      "Collecting jaxlib<=0.6.0,>=0.6.0 (from jax)\n",
      "  Downloading jaxlib-0.6.0-cp311-cp311-manylinux2014_x86_64.whl.metadata (1.2 kB)\n",
      "Collecting jax\n",
      "  Downloading jax-0.5.3-py3-none-any.whl.metadata (22 kB)\n",
      "Collecting jaxlib<=0.5.3,>=0.5.3 (from jax)\n",
      "  Downloading jaxlib-0.5.3-cp311-cp311-manylinux2014_x86_64.whl.metadata (1.2 kB)\n",
      "Collecting jax\n",
      "  Downloading jax-0.5.2-py3-none-any.whl.metadata (22 kB)\n",
      "Collecting jaxlib<=0.5.2,>=0.5.1 (from jax)\n",
      "  Downloading jaxlib-0.5.1-cp311-cp311-manylinux2014_x86_64.whl.metadata (978 bytes)\n",
      "Collecting jax\n",
      "  Downloading jax-0.5.1-py3-none-any.whl.metadata (22 kB)\n",
      "  Downloading jax-0.5.0-py3-none-any.whl.metadata (22 kB)\n",
      "Collecting jaxlib<=0.5.0,>=0.5.0 (from jax)\n",
      "  Downloading jaxlib-0.5.0-cp311-cp311-manylinux2014_x86_64.whl.metadata (978 bytes)\n",
      "Collecting jax\n",
      "  Downloading jax-0.4.38-py3-none-any.whl.metadata (22 kB)\n",
      "Collecting jaxlib<=0.4.38,>=0.4.38 (from jax)\n",
      "  Downloading jaxlib-0.4.38-cp311-cp311-manylinux2014_x86_64.whl.metadata (1.0 kB)\n",
      "Collecting opt_einsum (from jax)\n",
      "  Using cached opt_einsum-3.4.0-py3-none-any.whl.metadata (6.3 kB)\n",
      "Collecting multipledispatch (from numpyro)\n",
      "  Downloading multipledispatch-1.0.0-py3-none-any.whl.metadata (3.8 kB)\n",
      "Collecting tqdm (from numpyro)\n",
      "  Using cached tqdm-4.67.1-py3-none-any.whl.metadata (57 kB)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /opt/anaconda3/envs/L1EM/lib/python3.11/site-packages (from sympy->torch) (1.3.0)\n",
      "Downloading jax-0.4.38-py3-none-any.whl (2.2 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.2/2.2 MB\u001b[0m \u001b[31m18.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading jaxlib-0.4.38-cp311-cp311-manylinux2014_x86_64.whl (101.8 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m101.8/101.8 MB\u001b[0m \u001b[31m48.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m0:01\u001b[0m\n",
      "\u001b[?25hDownloading numpyro-0.18.0-py3-none-any.whl (365 kB)\n",
      "Downloading ml_dtypes-0.5.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.7 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.7/4.7 MB\u001b[0m \u001b[31m38.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading multipledispatch-1.0.0-py3-none-any.whl (12 kB)\n",
      "Using cached opt_einsum-3.4.0-py3-none-any.whl (71 kB)\n",
      "Using cached tqdm-4.67.1-py3-none-any.whl (78 kB)\n",
      "Installing collected packages: multipledispatch, tqdm, opt_einsum, ml_dtypes, jaxlib, jax, numpyro\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7/7\u001b[0m [numpyro]m6/7\u001b[0m [numpyro]s]\n",
      "\u001b[1A\u001b[2KSuccessfully installed jax-0.4.38 jaxlib-0.4.38 ml_dtypes-0.5.1 multipledispatch-1.0.0 numpyro-0.18.0 opt_einsum-3.4.0 tqdm-4.67.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install --upgrade torch git+https://github.com/itmat/BEERS2 jax numpyro\n",
    "%conda install pysam=0.23.3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f057ed0d",
   "metadata": {},
   "source": [
    "We then run the L1EM script as we made it with our changes on both practical and algorithmic ways."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63ff0d13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STEP 1: realign\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 0 reads\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -k 3 -n 3 -t 12 -i 20 /home/bio/git/l1emPipeline/data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa unaligned.fq1\n",
      "[main] Real time: 2.244 sec; CPU: 1.997 sec\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -k 3 -n 3 -t 12 -i 20 /home/bio/git/l1emPipeline/data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa unaligned.fq2\n",
      "[main] Real time: 2.511 sec; CPU: 1.998 sec\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa sampe /home/bio/git/l1emPipeline/data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa 1.sai 2.sai unaligned.fq1 unaligned.fq2\n",
      "[main] Real time: 0.005 sec; CPU: 0.002 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STEP 2: extract\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 89774 reads\n",
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 0 reads\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STEP 3: candidate alignments\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bwa_aln_core] calculate SA coordinate... 24.38 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq1.aa\n",
      "[main] Real time: 6.650 sec; CPU: 24.593 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 23.19 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq2.aa\n",
      "[main] Real time: 5.813 sec; CPU: 23.404 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 24.87 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq1.ab\n",
      "[main] Real time: 6.447 sec; CPU: 25.089 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 26.62 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq2.ab\n",
      "[main] Real time: 7.640 sec; CPU: 26.844 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 27.85 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq1.ac\n",
      "[main] Real time: 7.674 sec; CPU: 28.074 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 29.12 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq2.ac\n",
      "[main] Real time: 7.827 sec; CPU: 29.343 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 28.08 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq1.ad\n",
      "[main] Real time: 7.627 sec; CPU: 28.305 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 26.36 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq2.ad\n",
      "[main] Real time: 7.442 sec; CPU: 26.575 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 23.81 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq1.ae\n",
      "[main] Real time: 6.097 sec; CPU: 24.019 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 24.48 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq2.ae\n",
      "[main] Real time: 5.363 sec; CPU: 24.689 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 24.50 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq1.af\n",
      "[main] Real time: 4.734 sec; CPU: 24.708 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 24.90 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n",
      "[main] Version: 0.7.19-r1273\n",
      "[main] CMD: /opt/anaconda3/envs/L1EM/bin/bwa aln -t 12 -N -n 3 -k 3 -i 20 -R 10000000 /home/bio/git/l1emPipeline/L1EM/annotation/L1EM.400.ens.noY.fa L1.fq2.af\n",
      "[main] Real time: 4.567 sec; CPU: 25.104 sec\n",
      "[bwa_aln_core] calculate SA coordinate... 21.16 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3741 sequences have been processed.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "rm -rf ./output/L1EM\n",
    "# pushd ./L1EM/\n",
    "# bash -e ./L1EM/run_L1EM_3.sh ./data/experiment/SRR3997504/SRR3997504_sorted.bam ./L1EM ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa\n",
    "# popd\n",
    "# pushd ./L1EM\n",
    "# bash -e run_L1EM_3.sh /home/bio/git/l1emPipeline/output/beers2ashes/results/merged.sorted.bam /home/bio/git/l1emPipeline/L1EM /home/bio/git/l1emPipeline/data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa ./output/L1EM\n",
    "# popd\n",
    "bash -e ./L1EM/run_L1EM_3.sh ./output/beers2ashes/results/merged.sorted.bam ./L1EM ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa ./output/L1EM em\n",
    "# python ./L1EM/caclmetrics.py -ic ./output/run_1/CAMPAREE/data/sample1/transcript_quantifications.txt -il ./L1EM/full_counts.txt -o ./output/metrics.txt -otemp ./output/merged.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "347e863b",
   "metadata": {},
   "source": [
    "We finally calculate some metrics from the results of our custom L1EM script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21b41ae9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/bio/git/l1emPipeline/./src/calculateMetrics.py:53: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  df['family.category.locus.strand'] = df['family.category.locus.strand'].str.replace(\".\", \"_\")\n",
      "/home/bio/git/l1emPipeline/./src/calculateMetrics.py:56: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  df['family.category.locus.strand'] = df['family.category.locus.strand'].str.replace(\"+\", \"_\")\n",
      "/home/bio/git/l1emPipeline/./src/calculateMetrics.py:68: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  df['#transcript_id'] = df['#transcript_id'].str.replace('+','_')\n",
      "/home/bio/git/l1emPipeline/./src/calculateMetrics.py:69: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  df['#transcript_id'] = df['#transcript_id'].str.replace('.','_')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  locus  true_tpm         tpm\n",
      "10399  L1PA2_0_chr5_163684073_163684126       NaN  117.194125\n",
      "10400     L1HS_0_chr1_30567813_30568750       NaN   10.327296\n",
      "10401     L1HS_0_chr1_41207751_41208441       NaN   15.305655\n",
      "10402     L1HS_0_chr1_41854523_41855316       NaN   11.978680\n",
      "10403     L1HS_0_chr1_41989723_41990366       NaN   11.043149\n",
      "...                                 ...       ...         ...\n",
      "10698  L1PA2_0_chrX_131503357_131505329       NaN    3.503166\n",
      "10699  L1PA2_0_chrX_143003983_143004250       NaN    0.538070\n",
      "10700  L1PA2_0_chrX_150131407_150133104       NaN    0.863739\n",
      "10701   L1HS_0_chrX_152349286_152350533       NaN    5.511096\n",
      "10702  L1PA3_0_chrX_152386378_152387309       NaN    7.050222\n",
      "\n",
      "[304 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "mkdir -p ./output/metricsAshes\n",
    "rm -rf ./output/metricsAshes/*\n",
    "python ./src/calculateMetrics.py -ic ./output/run_1/CAMPAREE/data/sample1/transcript_quantifications.txt -il ./output/L1EM/full_counts.txt -o ./output/metricsAshes/metrics.txt -otemp ./output/metricsAshes/merged.txt -method EM"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd9d7260",
   "metadata": {},
   "source": [
    "We then calculate the L1EM results for the experiment SRR3997504 and its metrics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31814682",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "rm -rf ./output/L1EMSRR\n",
    "mkdir -p ./output/L1EMSRR\n",
    "bash -e ./L1EM/run_L1EM_3.sh ./data/experiment/SRR3997504_sorted.bam ./L1EM ./data/HomoSapiens_GRCh38_Ensemblv99/HomoSapiens_GRCh38_Ensemblv99.oneline_seqs.fa ./output/L1EMSRR vb\n",
    "mkdir -p ./output/metricsSRR\n",
    "rm -rf ./output/metricsSRR/*\n",
    "# python ./src/calculateMetrics.py -ic ./output/run_1/CAMPAREE/data/sample1/transcript_quantifications.txt -il ./output/L1EM/full_counts.txt -o ./output/metricsSRR/metrics.txt -otemp ./output/metricsSRR/merged.txt -method VB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "423bb011",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "L1EM",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
